dipy/denoise/localpca.py

```diff
@@ -0,0 +1,136 @@
+# on the standard scipy SVD API:
+except ImportError:
+    from scipy.linalg import svd
+    svd_args = [False]
+
+
+def localpca(arr, sigma, patch_radius=2, tau_factor=2.3, out_dtype=None):
+    r"""Local PCA-based denoising of diffusion datasets.
+
+    Parameters
+    ----------
+    arr : 4D array
+        Array of data to be denoised. The dimensions are (X, Y, Z, N), where N
+        are the diffusion gradient directions.
+    sigma : float or 3D array
+        Standard deviation of the noise estimated from the data.
+    patch_radius : int, optional
+        The radius of the local patch to be taken around each voxel (in
+        voxels). Default: 2 (denoise in blocks of 5x5x5 voxels).
+    tau_factor : float, optional
+        Thresholding of PCA eigenvalues is done by nulling out eigenvalues that
+        are smaller than:
+
+        .. math ::
+
+                \tau = (\tau_{factor} \sigma)^2
+
+        Default: 2.3, based on the results described in [Manjon13]_.
+    out_dtype : str or dtype, optional
+        The dtype for the output array. Default: output has the same dtype as
+        the input.
+
+    Returns
+    -------
+    denoised_arr : 4D array
+        This is the denoised array of the same size as that of the input data,
+        clipped to non-negative values
+
+    References
+    ----------
+    .. [Manjon13] Manjon JV, Coupe P, Concha L, Buades A, Collins DL (2013)
+                  Diffusion Weighted Image Denoising Using Overcomplete Local
+                  PCA. PLoS ONE 8(9): e73021.
+                  https://doi.org/10.1371/journal.pone.0073021
+    """
+    if out_dtype is None:
+        out_dtype = arr.dtype
+
+    # We retain float64 precision, iff the input is in this precision:
+    if arr.dtype == np.float64:
+        calc_dtype = np.float64
+
+    # Otherwise, we'll calculate things in float32 (saving memory)
+    else:
+        calc_dtype = np.float32
+
+    if not arr.ndim == 4:
+        raise ValueError("PCA denoising can only be performed on 4D arrays.",
+                         arr.shape)
+
+    patch_size = 2 * patch_radius + 1
+
+    if patch_size ** 3 < arr.shape[-1]:
+        e_s = "You asked for PCA denoising with a "
+        e_s += "patch_radius of {0} ".format(patch_radius)
+        e_s += "for data with {0} directions. ".format(arr.shape[-1])
+        e_s += "This would result in an ill-conditioned PCA matrix. "
+        e_s += "Please increase the patch_radius."
+        raise ValueError(e_s)
+
+    if isinstance(sigma, np.ndarray):
+        if not sigma.shape == arr.shape[:-1]:
+            e_s = "You provided a sigma array with a shape"
+            e_s += "{0} for data with".format(sigma.shape)
+            e_s += "shape {0}. Please provide a sigma array".format(arr.shape)
+            e_s += " that matches the spatial dimensions of the data."
+            raise ValueError(e_s)
+
+    tau = np.median(np.ones(arr.shape[:-1]) * ((tau_factor * sigma) ** 2))
```

Comment:
@arokem Sorry, I don't understand why you are using the median of Sigma, instead of the whole array. I couldn't find it in the paper. Isn't it more accurate to use the whole array, which has a specific value for each voxel?